modules/client-handler/src/main/java/org/apache/ignite/client/handler/JdbcQueryEventHandlerImpl.java

```diff
@@ -164,8 +175,58 @@ public JdbcQueryEventHandlerImpl(QueryProcessor processor, JdbcMetadataCatalog m
     /** {@inheritDoc} */
     @Override
     public CompletableFuture<BatchExecuteResult> batchAsync(BatchExecuteRequest req) {
-        return CompletableFuture.completedFuture(new BatchExecuteResult(UNSUPPORTED_OPERATION,
-                "ExecuteBatch operation is not implemented yet."));
+        IntList res = new IntArrayList();
+
+        IgniteBiTuple<Integer, String> firstError = new IgniteBiTuple<>();
+        try {
+            for (Query query : req.queries()) {
+                if (query.args() == null) {
+                    executeAndCollectUpdateResult(req, res, query, OBJECT_EMPTY_ARRAY, firstError);
+                } else {
+                    for (Object[] arg : query.args()) {
+                        executeAndCollectUpdateResult(req, res, query, arg, firstError);
+                    }
+                }
+            }
+        } catch (Exception e) {
+            StringWriter sw = getWriterWithStackTrace(e);
+
+            return CompletableFuture.completedFuture(new BatchExecuteResult(Response.STATUS_FAILED,
+                    "Exception while executing query " + req.queries().get(0).sql() + ". Error message: " + sw));
+        }
+
+        if (!firstError.isEmpty()) {
+            return CompletableFuture.completedFuture(new BatchExecuteResult(
```

Comment:
How to understand what query from the batch has failed if the 'res' contains results for all of the queries?

Reply:
The array element for any command that failed is Statement.EXECUTE_FAILED (-3).
On the other hand, there can only be one error, and JDBC driver contract does not regulate it for the case when several queries are failed.

My implementation is the same as the one in Ignite 2: only the error of the first broken query is sent to the client. The rest of the errors are logged on the server node. 
Maybe we should discuss this behavior.

Reply:
I refactored the exception handling. We won't continue with operations if the execution of a query ended with an error. Now, the result will always contain only positive results of update operations.

updateCounts – an update count or Statement.SUCCESS_NO_INFO for each SQL command in the batch prior to the failure for JDBC drivers that stop processing after a command failure